pkg/apis/policy/v1alpha1/upstreamtraffic.go

```diff
@@ -0,0 +1,103 @@
+type UpstreamTraffic struct {
+	// Object's type metadata
+	metav1.TypeMeta `json:",inline"`
+
+	// Object's metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the UpstreamTraffic policy specification
+	// +optional
+	Spec UpstreamTrafficSpec `json:"spec,omitempty"`
+}
+
+// UpstreamTrafficSpec defines the upstream traffic policy specification.
+type UpstreamTrafficSpec struct {
+	// Host the upstream traffic is directed to.
+	// Must either be an FQDN corresponding to the upstream service
+	// or the name of the upstream service. If only the service name
+	// is specified, the FQDN is derived from the service name and
+	// the namespace of the UpstreamTraffic rule.
+	Host string `json:"host"`
+
+	// ConnectionSettings specifies the connection settings for traffic
+	// directed to the upstream host.
+	// +optional
+	ConnectionSettings ConnectionSettingsSpec `json:"connectionSettings,omitempty"`
+}
+
+// ConnectionSettingsSpec defines the connection settings for an
+// upstream host.
+type ConnectionSettingsSpec struct {
+	// TCP specifies the TCP level connection settings.
+	// Applies to both TCP and HTTP connections.
+	// +optional
+	TCP *TCPConnectionSettings `json:"tcp,omitempty"`
+
+	// HTTP specifies the HTTP level connection settings.
+	// +optional
+	HTTP *HTTPConnectionSettings `json:"http,omitempty"`
+}
+
+// TCPConnectionSettings defines the TCP connection settings for an
+// upstream host.
+type TCPConnectionSettings struct {
+	// MaxConnections specifies the maximum number of TCP connections
+	// allowed to the upstream host.
+	// Defaults to 1024 if not specified.
+	// +optional
+	MaxConnections uint32 `json:"maxConnections,omitempty"`
+
+	// ConnectTimeout specifies the TCP connection timeout.
+	// Defaults to 5s if not specified.
+	// +optional
+	ConnectTimeout metav1.Duration `json:"connectTimeout,omitempty"`
+}
+
+// HTTPConnectionSettings defines the HTTP connection settings for an
+// upstream host.
+type HTTPConnectionSettings struct {
+	// MaxRequests specifies the maximum number of parallel requests
+	// allowed to the upstream host.
+	// Defaults to 1024 if not specified.
+	// +optional
+	MaxRequests uint32 `json:"maxRequests,omitempty"`
+
+	// MaxRequestsPerConnection specifies the maximum number of requests
+	// per connection allowed to the upstream host.
+	// Defaults to unlimited if not specified.
+	// +optional
+	MaxRequestsPerConnection uint32 `json:"maxRequestsPerConnection,omitempty"`
+
+	// MaxPendingRequests specifies the maximum number of pending HTTP/1.1
+	// requests allowed to the upstream host.
+	// Defaults to 1024 if not specified.
+	// +optional
+	MaxPendingRequests uint32 `json:"maxPendingRequests,omitempty"`
+
+	// MaxRetries specifies the maximum number of parallel retries
+	// allowed to the upstream host.
+	// Defaults to 3 if not specified.
+	// +optional
+	MaxRetries uint32 `json:"maxRetries,omitempty"`
```

Comment:
To Jon's point, wouldn't 0 be a valid value for MaxRetries? If that's the case then the statement "Defaults to 3 if not specified" is incorrect, since we can't distinguish between 0 and not specified

Reply:
At this point, there is no implementation backing the API to render defaults. The controller would have logic like
`if (maxRetries > 0) // configure max retries`. I don't yet know whether this can be set to zero, needs testing, so at this point made the assumption that a non-zero value will not be allowed for max_retries similar to the other uint32 values.